internal/controller/bmcutils.go

```diff
@@ -15,15 +15,37 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-func GetBMCClientFromBMCName(ctx context.Context, c client.Client, bmcName string, insecure bool) (bmc.BMC, error) {
-	bmc := &metalv1alpha1.BMC{}
-	if err := c.Get(ctx, client.ObjectKey{Name: bmcName}, bmc); err != nil {
-		if errors.IsNotFound(err) {
-			return nil, fmt.Errorf("BMC %q not found", bmcName)
+func GetBMCClientForServer(ctx context.Context, c client.Client, server *metalv1alpha1.Server, insecure bool) (bmc.BMC, error) {
+	if server.Spec.BMCRef != nil {
+		b := &metalv1alpha1.BMC{}
+		bmcName := server.Spec.BMCRef.Name
+		if err := c.Get(ctx, client.ObjectKey{Name: bmcName}, b); err != nil {
+			if errors.IsNotFound(err) {
+				return nil, fmt.Errorf("BMC %q not found", bmcName)
+			}
+			return nil, err
 		}
-		return nil, err
+
+		return GetBMCClientFromBMC(ctx, c, b, insecure)
+	}
+
+	if server.Spec.BMC != nil {
+		bmcSecret := &metalv1alpha1.BMCSecret{}
+		if err := c.Get(ctx, client.ObjectKey{Name: server.Spec.BMC.BMCSecretRef.Name}, bmcSecret); err != nil {
+			return nil, fmt.Errorf("failed to get BMC secret: %w", err)
```

Comment:
Should we handle the NotFound error gracefully here, just to avoid plenty of reconciles in case the object doesn't exist?

Reply:
We need to fail fast here as we can't construct the BMC client if we can't get the credentials.